### ISSUE TITLE
RFC: non-owning filter

### ISSUE THREAD

**Issue by hellerve** [carp-lang/Carp#317]
The current implementation of `filter` takes ownership of the array it’s being passed. This makes sense, because it modifies the array in-place, but it also means that we always have to copy the array before passing it in if we want to keep working with it, which in my experience is a fairly common workflow.

To make this easier and more performant, I propose that we add a function `filter-copy`, similar to `copy-map`, that copies the array. Whether it should be called `filter-copy` or `copy-filter` I’m not entirely sure about. Aesthetically I prefer the former, but the latter is more consistent (and consistency is what’s ultimately more important).

This needs to be added as an array template, and I’d be up to add this if we reach a consensus whether this makes sense or just adds bloat.

**Comment by eriksvedang** [comment]
I think it's a fairly common use case and in some cases this optimisation will mean a lot (especially if a big array is filtered down to something small). Consistent naming using `copy-filter` seems like the best option, yes. And even though none of the names are particularly beautiful, at least they warn the programmer about their copying nature, which is also good.

**Comment by hellerve** [comment]
Should `filter` then be renamed to `endo-filter`?

**Comment by chrisosaurus** [comment]
+1 to being explicit about copys, it is one of the biggest problems I often face in C++ (accidental copies are too easy).

**Comment by hellerve** [comment]
I think we needn’t be consistent, then!

**Comment by hellerve** [comment]
There is a tradeoff about the characteristics of `copy-filter`, though. It could look like this:

```clojure
(defmodule Array
  (defn copy-filter [f a]
    (let-do [res []]
      (for [i 0 (length a)]
        (when (f @(nth a i)) (push-back! &res @(nth a i))))
      res))
)
```

This would have to allocate every once in a while, but would be good if we filter “aggressively”, i.e. if only few elements remain. On the other hand, if we implemented it like this:

```C
(defmodule Array
  (defn copy-filter [f a] (filter f @a))
)
```

It would copy the whole array once, and thus perform less allocations; it would, however, waste space. This would go away if we optimized `filter` as planned, however (resizing it properly after we’re done filtering).

What do y’all think?
